pc2ccs/pc2v9
# I538 support interactive problems

### Description of what the PR does
Add built-in support for interactive problems.
Previously, interactive problems were handled by munging the execute command line from the edit problem GUI or in a problem's YAML file.
This PR makes PC2 aware of interactive problems.  Special custom shell scripts are used to setup the pipe work needed to connect a run's I/O to the O/I of a CLICS compliant custom interactive validator.  There are separate scripts for sandbox and non-sandbox problems.
An additional special script is used during the PC2 `validateProgram()` phase that copies the results of an interactive run performed by one of the aforementioned shell scripts to the location desired by PC2.

Summary of changes to the code:

- Add support in `ContestSnakeYAMLLoader.java` to recognize an interactive problem and perform appropriate validation of the validator supplied.
- Add support to the `ProblemPane` GUI to indicate if a problem is interactive.
- Add support to the `EditProblemPane `GUI to allow configuration of interactive problems.
- Add support for interactive problems to the `Problem` object.
- Expandthe `CustomValidatorSettings `object to be aware of interactive problems.
- `Executable.java`

  - Change to deal with interactive problems.
  - Refactor some code to put commonly used blocks in methods.
  - Add more substitution variables to support interactive problems and also allow for better logging in validators (`{:testcase}`, `{:executeinfofilename}`) .
  - Add routines to set up the interactive environment in the execute folder.
  - Modify build.xml and package.xml to correctly Zip and (gzip) the new **scripts** folder.  Remove  the **sandbox** folder.

- Create new scripts folder to hold all helper scripts used by sandboxes and interactive validators.  This involved getting rid of the `sandbox `folder (really, renaming it) and creating the `scripts `folder. 
- CI: Fix missing "`toString`" properties.
- CI: Make` teams2.tsv` be the preferred teams file during loading of a CDP.  This provides organizations.
- CI: better reporting of errors during reading of the CDP when loading a contest.  Combats the antagonizing "_check logs_" message previously reported.
- CI: Update copyright notices

There is still some future work to do in reading the results of the interactive validator (and sandbox) and reporting them on the GUI in PC2.  The current interactive and sandbox scripts create logs and "`reports`" in the "`execute*`" folder.  These could be read by PC2 (in the future) and reported to the user.  In addition, the current report files should be changed to create a single **NDJSON** type file (the framework is already in place for this).  This is NOT required for correct operation of interactive problems.  It would, however, make it nicer for someone reviewing results or manually judging.

### Issue which the PR addresses
Fixes #538

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11/Ubuntu 22.04.2
Eclipse
java version "1.8.0_321"
Java(TM) SE Runtime Environment (build 1.8.0_321-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.321-b07, mixed mode)

### Precise steps for _testing_ the PR 
**I might suggest TESTING the PR first before doing a code review so we can shake out any bugs before the code review.**

Please allocate enough time and patience to test this PR.  It's not a 10 minute task.  **This only works on Linux** due to sandboxes and interactive problems being used.  Appropriate warnings and denials will occur on _Windows_.  In addition, it requires the use of cgoupv2 - so be sure it's configured on the Linux system you are testing on.  The scripts including with PC2 will setup and configure the cgroup required by PC2, assuming the kernel supports it.

1. Load the `samps/contest/interactive` contest.  This is a new sample contest and it has 2 interactive problems (**Amazing** and **guess**), and a non-interactive problem (**pollution solution**)
2. By default, the CDP arranges for all problems to be judged by **AJ Judge1**.  If you want to manually judge, then you'll have to modify the problem and set it to manual judging.
3. Start an admin client.
4. Set the **Judges Data Path** so it points to the correct spot, namely:  basedir/samps/contests/interactive/config (or wherever)
5. Start **judge1** client (the autojudge) - if you get errors about missing data, then your **Judge Data Path** is wrong.
6. Start the contest (if you didn't use the GUI server, you'll need to start an admin client to start the contest)
7. Start **team1** client.
8. Submit one of the accepted solutions to **Amazing** (`samps/interactive/config/amazing/submissions/accepted`)
9. Observe AJ behavior and result received by team client.  Should be **correct**.
10. Submit more runs from **Amazing** and **Guess** to confirm interactive operation works.
11. Submit one or more runs from the **Pollution** problem.  This is a standard, non-interactive, built-in _CLICS_ validator problem (consider it a "_control_" problem).
12. Now submit some of the Wrong answer or time limit exceeded submissions also found under each problem.
13. Confirm correct operation.

Congratulations, you've confirmed that the PR works (assuming, it did!)

Now, do the code review.  First, calm yourself... even though there appear to be lots of commits, many of them are merge's from **develop** to keep the code up-to-date.  Other commits are for comments.  Others are because things were added, like the `samps/contests/interactive contest`.  This is not to say there aren't a lot of changes: there are.  It's just not as overwhelming as it may seem based on the number of commits and _Files Changed_ (currently, a whopping 212 files - mostly new ones for the new sample contest - so again, relax).
